### PR TITLE
Add support for multi-level includes

### PR DESCRIPTION
**Pull request by tylermmorton**
This PR is in response to #390 -- This adds support for multi-level includes by rewiring `read.Taskfile` to be a recursive function that merges Taskfiles in depth first order. It also uses the new `read.ReaderNode` to keep track of included Taskfile parent nodes and checks for include cycles that could cause overflow issues.

I've also added some basic tests to confirm the new functionality. Happy to continue to refine this if any interest is shown in the feature. This would absolutely be a game changer for a lot of the projects I'm involved in.

### CODE REVIEW

File: docs/usage.md [hunk 1]
```diff
@@ -8,7 +8,7 @@ The example below allows compiling a Go app and uses [Minify][minify] to concat
 and minify multiple CSS files into a single one.
 
 ```yaml
-version: '3'
+version: "3"
```
Comment (andreynering) [inline docs/usage.md:11]: This was changed this the whole file, but single quotes are really preferred according to out [styleguide](https://taskfile.dev/#/styleguide).

File: taskfile/read/taskfile.go [hunk 2]
```diff
@@ -170,3 +179,25 @@ func exists(path string) (string, error) {
 
 	return "", fmt.Errorf(`task: No Taskfile found in "%s". Use "task --init" to create a new one`, path)
 }
+
+func checkCircularIncludes(node *ReaderNode) error {
+	if node == nil {
+		return errors.New("failed to check for include cycle: node was nil")
```
Comment (andreynering) [inline taskfile/read/taskfile.go:185]: We prefix all errors with `task: `.

```suggestion
		return errors.New("task: failed to check for include cycle: node was nil")
```

File: taskfile/read/taskfile.go [hunk 3]
```diff
@@ -170,3 +179,25 @@ func exists(path string) (string, error) {
...
+	}
+	if node.Parent == nil {
+		return errors.New("failed to check for include cycle: node.Parent was nil")
```
Comment (andreynering) [inline taskfile/read/taskfile.go:188]:
```suggestion
		return errors.New("task: failed to check for include cycle: node.Parent was nil")
```

File: taskfile/read/taskfile.go [hunk 4]
```diff
@@ -170,3 +179,25 @@ func exists(path string) (string, error) {
...
+	}
+	var curNode = node
+	var basePath = filepath.Join(node.Dir, node.Entrypoint)
+	for curNode.Parent != nil {
+		curNode = curNode.Parent
+		curPath := filepath.Join(curNode.Dir, curNode.Entrypoint)
+		if curPath == basePath {
+			return fmt.Errorf("include cycle detected between %s <--> %s",
```
Comment (andreynering) [inline taskfile/read/taskfile.go:196]:
```suggestion
			return fmt.Errorf("task: include cycle detected between %s <--> %s",
```

File: testdata/includes_cycle/Taskfile.yml [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+includes:
+  'one': ./one/Taskfile.yml
+
+tasks:
+  default:
+    cmds:
+      - echo "called_dep" > called_dep.txt
+  level1:
+    cmds:
+      - echo "hello level 1"
```
Comment (andreynering) [inline testdata/includes_cycle/Taskfile.yml:12]: We need to add a final newline to all these test files.

**Review by andreynering — COMMENTED** [review]
Hi @tylermmorton,

Thanks again for yet another PR!

I haven't tested this one yet, but the code looks good, and I added a few comments for now. Hope I'll find the time to test it soon.

File: docs/usage.md (hunk 1 above)
Comment (tylermmorton) [reply docs/usage.md]: Fixed!

File: taskfile/read/taskfile.go [hunk 6]
```diff
@@ -68,24 +74,27 @@ func Taskfile(dir string, entrypoint string) (*taskfile.Taskfile, error) {
...
 		}
 
-		path, err = exists(path)
-		if err != nil {
-			if includedTask.Optional {
-				return nil
-			}
+		includeReaderNode := &ReaderNode{
+			Dir:        filepath.Dir(path),
+			Entrypoint: filepath.Base(path),
```
Comment (andreynering) [inline taskfile/read/taskfile.go:82]: Seems that we have the wrong value here when inclusion is by a directory (omits `Taskfile.yml`).

**Review by andreynering — COMMENTED** [review]
Hi @tylermmorton,

The code looks great, I think this is the right direction. I really appreciate your effort.

I did some basic testing and found a small bug: when you do the inclusion like `foo: ./foo` instead of `foo: ./foo/Taskfile.yml` you get an error.

Reproduce:

- I added a new Taskfile on `docs/foo/Taskfile.yml` as:

```yaml
version: '3'

tasks:
  echo:
    desc: Includes inside includes
    cmds:
      - echo "Includes inside includes"
      - pwd
```

- Referenced it from the `docs/Taskfile.yml` like:

```yaml
includes:
  foo: ./foo
```

And the error is like: `stat /Users/andrey/Developer/andrey/task/foo: no such file or directory`. Note the dir is `task/foo` instead of `task/docs/foo`.

I think the issue is the place pointed in a review comment below. We should have a test case covering this.

The directory where the task _runs_ should also be tested, but this one seems to be working fine in my tests (once `/Taskfile.yml` was added to includes due to the previous issue).

File: taskfile/read/taskfile.go (hunk 6 above)
Comment (tylermmorton) [reply taskfile/read/taskfile.go:89]: Fixed! I added a check for `exists()` so all default taskfile names can be used 😄

**Comment by lukepighetti** [comment]
Hate to be that guy, but my experience is that big lift PRs that sit stale never land. We're looking for this feature and would love to see this merged.

**Review by andreynering — APPROVED** [review]

**Comment by andreynering** [comment]
Thanks again @tylermmorton!

**Pull request merged**